transport/x/grpc/inbound.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"errors"
+	"net"
+	"sync"
+
+	"go.uber.org/yarpc/api/transport"
+	internalsync "go.uber.org/yarpc/internal/sync"
+
+	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
+	"google.golang.org/grpc"
+)
+
+var (
+	errRouterNotSet          = errors.New("router not set")
+	errRouterHasNoProcedures = errors.New("router has no procedures")
+
+	_ transport.Inbound = (*Inbound)(nil)
+)
+
+// Inbound is a grpc transport.Inbound.
+type Inbound struct {
+	once           internalsync.LifecycleOnce
+	lock           sync.Mutex
+	listener       net.Listener
+	inboundOptions *inboundOptions
+	router         transport.Router
+	server         *grpc.Server
+}
+
+// NewInbound returns a new Inbound for the given listener.
+func NewInbound(listener net.Listener, options ...InboundOption) *Inbound {
+	return &Inbound{internalsync.Once(), sync.Mutex{}, listener, newInboundOptions(options), nil, nil}
+}
+
+// Start implements transport.Lifecycle#Start.
+func (i *Inbound) Start() error {
+	return i.once.Start(i.start)
+}
+
+// Stop implements transport.Lifecycle#Stop.
+func (i *Inbound) Stop() error {
+	return i.once.Stop(i.stop)
+}
+
+// IsRunning implements transport.Lifecycle#IsRunning.
+func (i *Inbound) IsRunning() bool {
+	return i.once.IsRunning()
+}
+
+// SetRouter implements transport.Inbound#SetRouter.
+func (i *Inbound) SetRouter(router transport.Router) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.router = router
+}
+
+// Transports implements transport.Inbound#Transports.
+func (i *Inbound) Transports() []transport.Transport {
+	return []transport.Transport{}
+}
+
+func (i *Inbound) start() error {
+	i.lock.Lock()
+	defer i.lock.Unlock()
```

Comment:
instead of locking the whole start function, we should make a function "getProcedures" that locks around the router and returns the procedures from the route (or an error).